src/psyclone/domain/nemo/transformations/nemo_arrayaccess2loop_trans.py

```diff
@@ -0,0 +1,328 @@
+            # kernel.
+            CreateNemoKernelTrans().apply(assignment.parent)
+
+    def validate(self, node, options=None):
+        '''Perform various checks to ensure that it is valid to apply the
+        NemoArrayAccess2LoopTrans transformation to the supplied PSyIR Node.
+
+        :param node: the node that is being checked.
+        :type node: :py:class:`psyclone.psyir.nodes.Node`
+        :param options: a dictionary with options for \
+            transformations. No options are used in this \
+            transformation. This is an optional argument that defaults \
+            to None.
+        :type options: dict of string:values or None
+
+        '''
+        # Not a PSyIR node
+        if not isinstance(node, Node):
+            raise TransformationError(
+                "Error in NemoArrayAccess2LoopTrans transformation. The "
+                "supplied node argument should be a PSyIR Node, but found "
+                "'{0}'.".format(type(node).__name__))
+        # Not within an array reference
+        if not node.parent or not isinstance(node.parent, ArrayReference):
+            raise TransformationError(
+                "Error in NemoArrayAccess2LoopTrans transformation. The "
+                "supplied node argument should be within an ArrayReference "
+                "node, but found '{0}'.".format(type(node.parent).__name__))
+        array_ref = node.parent
+        # Array reference not within an assignment
+        if not array_ref.parent or not isinstance(array_ref.parent,
+                                                  Assignment):
+            raise TransformationError(
+                "Error in NemoArrayAccess2LoopTrans transformation. The "
+                "supplied node argument should be within an ArrayReference "
+                "node that is within an Assignment node, but found '{0}'."
+                .format(type(array_ref.parent).__name__))
+        assignment = array_ref.parent
+        # Array reference not on lhs of the assignment
+        if assignment.lhs is not array_ref:
+            raise TransformationError(
+                "Error in NemoArrayAccess2LoopTrans transformation. The "
+                "supplied node argument should be within an ArrayReference "
+                "node that is within the left-hand-side of an Assignment "
+                "node, but it is on the right-hand-side.")
+
+        # Contains a range node
+        if node.walk(Range):
+            raise TransformationError(
+                "Error in NemoArrayAccess2LoopTrans transformation. The "
+                "supplied node should not be or contain a Range node as it "
+                "should be single valued.")
+
+        # Capture loop iterator symbols in order
+        iterator_symbols = []
+        location = node.parent.parent
+        while (isinstance(location.parent, Schedule) and
+               isinstance(location.parent.parent, (Loop, NemoKern))):
```

Comment:
`NemoKern`  complicates the use of `ancestor`?

Reply:
This code has been removed, but the general issue remains. The code may, or may not be part of a NemoKern. If it is we need to place the loop outside of NemoKern. If one does not exist we may need to add one and if we do we again need to place the loop outside of NemoKern.

Reply:
Ah, I understand now.